TST: add unit tests for mean, stddev, and variance for some

rlucas7 opened the pull request:
*Issue #, if available:*
Unit tests for distributions methods, gh-issue #46
*Description of changes:*
Adds unit tests for `mean`, `stddev`, and `variance` methods of `distribution` classes. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

rlucas7 commented:
@lostella here is a first pass, this doesn't have all the distributions, nor all the methods in gh-46 but it does have a general approach to getting all the methods and functionalities of `distribution` class. Opening this PR sooner rather than later in case there is a different preferred approach.

szha commented:
Job PR-182/1 is complete. 
Docs are uploaded to http://gluon-ts-staging.s3-accelerate.dualstack.amazonaws.com/PR-182/1/index.html

geoalgo commented on test/distribution/test_distribution_methods.py:
```diff
@@ -0,0 +1,107 @@
+    },
+}
+
+# TODO: work out all means, stddevs, variances as similar to test_cases structure
```
Could you add the list of distribution that are missing this test? MultivariateGaussian, LowrankMultivariateGaussian and Uniform comes to mind for instance.
```suggestion
# TODO: work out all means, stddevs, variances as similar to test_cases structure. Missing cases includes MultivariateGaussian, LowrankMultivariateGaussian and Uniform.
```

geoalgo commented on .gitignore:
```diff
@@ -159,3 +159,7 @@ conda
 # vscode
 .vscode
+
+#file in my repo says don't add to VCS
```
Could you remove this comment 'my repo' is weird in an open-source project ;-)

geoalgo replied:
```suggestion
```

lostella commented:
Looks like `black` is complaining: try running
```
black src test examples
```
and committing the proposed reformatting.

rlucas7 commented:
> Looks like `black` is complaining: try running
> 
> ```
> black src test examples
> ```
> 
> and committing the proposed reformatting.

Does it make sense that a bunch of files may be reformatted even if I didn't make any changes? 
Here is what black modifies:

```
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   src/gluonts/distribution/piecewise_linear.py
	modified:   src/gluonts/distribution/transformed_distribution_output.py
	modified:   test/distribution/test_distribution_inference.py
	modified:   test/distribution/test_distribution_methods.py
	modified:   test/distribution/test_transformed_distribution.py
```

rlucas7 commented:
@geoalgo The MultivariateGaussian throws an error for `stddev` method so I cannot do that one for now. 

```
(gluon-ts) Lucass-MacBook:gluon-ts rlucas$ python
Python 3.6.8 |Anaconda, Inc.| (default, Dec 29 2018, 19:04:46) 
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> # now for Multivariate Gaussian
... from gluonts.distribution import MultivariateGaussian
>>> import mxnet as mx
>>> import numpy as np
>>> lower=mx.nd.array([[1, 0, 0],[0, 1, 0], [0, 0, 1]], dtype='float32')
>>> mg=MultivariateGaussian(mu=mx.nd.zeros(shape=(3, 3)), L=lower)
>>> mg.mean

[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]
<NDArray 3x3 @cpu(0)>
>>> mg.variance

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
<NDArray 3x3 @cpu(0)>
>>> mg.stddev
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/rlucas/gluon-ts/src/gluonts/distribution/distribution.py", line 211, in stddev
    raise NotImplementedError()
NotImplementedError
>>> 
```

I'll add a TODO comment for that one.

szha commented:
Job PR-182/2 is complete. 
Docs are uploaded to http://gluon-ts-staging.s3-accelerate.dualstack.amazonaws.com/PR-182/2/index.html

rlucas7 commented:
Not sure what the failures are about, the unit tests all pass locally. 
Also, `black` is now happy. 
```
(gluon-ts) Lucass-MacBook:gluon-ts rlucas$ black src test examples
All done! ✨ 🍰 ✨
180 files left unchanged.
```

lostella commented:
I pulled your branch and black had complaints on my machine: I pushed the changes, let's see how it goes

szha commented:
Job PR-182/3 is complete. 
Docs are uploaded to http://gluon-ts-staging.s3-accelerate.dualstack.amazonaws.com/PR-182/3/index.html

jaheba commented:
@rlucas7 It might be that you use a different version of black. I think we are currently using `19.3b0`.

Can you check by running `pip show black`?

rlucas7 commented:
@jaheba Yeah for some reason I had an older version, was the requirements file updated? 

`Uninstalling black-18.6b4:` and then `pip install black` put in `Successfully installed black-19.3b0`
and then when I reran on this branch w/new black version I get: 

```
(gluon-ts) Lucass-MacBook:gluon-ts rlucas$ black src test examples
reformatted /Users/rlucas/gluon-ts/src/gluonts/distribution/transformed_distribution_output.py
reformatted /Users/rlucas/gluon-ts/src/gluonts/distribution/piecewise_linear.py
reformatted /Users/rlucas/gluon-ts/test/distribution/test_transformed_distribution.py
reformatted /Users/rlucas/gluon-ts/test/distribution/test_distribution_methods.py
reformatted /Users/rlucas/gluon-ts/test/distribution/test_distribution_inference.py
All done! ✨ 🍰 ✨
5 files reformatted, 175 files left unchanged.
```
which is (I'm guessing) what @lostella found.

szha commented:
Job PR-182/4 is complete. 
Docs are uploaded to http://gluon-ts-staging.s3-accelerate.dualstack.amazonaws.com/PR-182/4/index.html

lostella commented on test/distribution/test_distribution_methods.py:
```diff
@@ -0,0 +1,142 @@
+    assert means.shape == test_output[distr_name]['mean'].shape
+    # asnumpy()  needed to b/c means is all pointers to values
+    assert np.allclose(
+        means.asnumpy(), test_output[distr_name]['mean'].asnumpy(), atol=1e-2
```
This value for `atol` seems a bit too large. For example, in some assertions we are comparing against rather small values (around `1e-4` in one case). What happens if you leave `atol` to its default value in all assertions?

rlucas7 replied:
You can make them [go to 11](https://www.youtube.com/watch?v=4xgx4k83zzc) and all tests pass:
```
(gluon-ts) Lucass-MacBook:gluon-ts rlucas$ git diff --staged
diff --git a/test/distribution/test_distribution_methods.py b/test/distribution/test_distribution_methods.py
index 3c9f283..18fc80f 100644
--- a/test/distribution/test_distribution_methods.py
+++ b/test/distribution/test_distribution_methods.py
@@ -112,7 +112,7 @@ def test_means(distr_class, params) -> None:
     assert means.shape == test_output[distr_name]['mean'].shape
     # asnumpy()  needed to b/c means is all pointers to values
     assert np.allclose(
-        means.asnumpy(), test_output[distr_name]['mean'].asnumpy(), atol=1e-2
+        means.asnumpy(), test_output[distr_name]['mean'].asnumpy(), atol=1e-11
     )
 
 
@@ -125,7 +125,7 @@ def test_stdevs(distr_class, params) -> None:
     assert np.allclose(
         stddevs.asnumpy(),
         test_output[distr_name]['stddev'].asnumpy(),
-        atol=1e-2,
+        atol=1e-11,
     )
 
 
@@ -138,5 +138,5 @@ def test_variances(distr_class, params) -> None:
     assert np.allclose(
         variances.asnumpy(),
         test_output[distr_name]['variance'].asnumpy(),
-        atol=1e-2,
+        atol=1e-11,
     )
(gluon-ts) Lucass-MacBook:gluon-ts rlucas$  python -m pytest test/distribution/test_distribution_methods.py 
Setting module np/mx/python random seeds, use MXNET_MODULE_SEED=556451365 to reproduce.
================================================================================ test session starts ================================================================================
platform darwin -- Python 3.6.8, pytest-3.10.1, py-1.8.0, pluggy-0.9.0
rootdir: /Users/rlucas/gluon-ts, inifile: pytest.ini
plugins: xdist-1.27.0, timeout-1.2.1, remotedata-0.3.1, openfiles-0.3.2, forked-1.0.2, doctestplus-0.3.0, cov-2.6.1, arraydiff-0.3
collected 18 items                                                                                                                                                                  

test/distribution/test_distribution_methods.py ..................                                                                                                             [100%]

============================================================================= 18 passed in 0.16 seconds =============================================================================
(gluon-ts) Lucass-MacBook:gluon-ts rlucas$ 
```

rlucas7 replied:
the `atol=1e-2` were inherited from `test/distribution/test_distribution_sampling.py` when I created a new file for testing I `cp`-ed that one.

szha commented:
Job PR-182/5 is complete. 
Docs are uploaded to http://gluon-ts-staging.s3-accelerate.dualstack.amazonaws.com/PR-182/5/index.html

lostella commented:
@rlucas7 thank you! one check is failing because of a flaky test, but this PR is safe to merge
